plutus-core/plutus-ir/src/PlutusIR/Transform/Inline.hs

```diff
@@ -250,20 +262,41 @@ processSingleBinding = \case
     b -> Just <$> forMOf bindingSubterms b processTerm
 
 maybeAddSubst
-    :: Inlining tyname name uni fun a m
+    :: forall tyname name uni fun a m . Inlining tyname name uni fun a m
     => Strictness
     -> name
     -> Term tyname name uni fun a
     -> m (Maybe (Term tyname name uni fun a))
 maybeAddSubst s n rhs = do
-    -- Only do PostInlineUnconditional
-    -- See Note [Inlining approach and 'Secrets of the GHC Inliner']
-    rhs' <- processTerm rhs
-    doInline <- postInlineUnconditional s rhs'
-    if doInline then do
-        modify (\subst -> extendTerm n (Done rhs') subst)
-        pure Nothing
-    else pure $ Just rhs'
+    -- NOTE:  Nothing means that we are inlining the term:
+    --   * we have extended the substitution, and
+    --   * we are removing the binding (hence we return Nothing)
+    preUnconditional <- preInlineUnconditional n rhs
+    if preUnconditional then do
```

Comment:
The chain of logic here feels a little clunky, but I don't have a better suggestion. My best idea was to run this in `MaybeT m` so that it would short-circuit the first time you hit a `Nothing` indicating that we're inlining the binding... but that's quite obscure!

Reply:
I was actually thinking about using either `Alternative` or `MonadPlus`, because I think I saw somewhere in the current codebase short-circuiting versions of combinators on them.

Ideally we need  something similar to [cond](https://hackage.haskell.org/package/cond-0.4.1.1/docs/Control-Conditional.html), where we can specify multiple branches and predicates.

Reply:
I think it's fine as it is for now!